contracts/TokenGovernance.sol

```diff
@@ -0,0 +1,149 @@
+
+/**
+ * Copyright (c) 2018-present, Leap DAO (leapdao.org)
+ *
+ * This source code is licensed under the Mozilla Public License, version 2,
+ * found in the LICENSE file in the root directory of this source tree.
+ */
+
+pragma solidity 0.5.2;
+
+import "../node_modules/openzeppelin-solidity/contracts/token/ERC20/IERC20.sol";
+import "./Bridge.sol";
+import "./Vault.sol";
+import "./TxLib.sol";
+
+contract TokenGovernance {
+
+  uint256 constant PROPOSAL_STAKE = 5000000000000000000000;
+  uint32 constant PROPOSAL_TIME = 604800; // 60 * 60 * 24 * 7 = 7 days
+  address public mvgAddress;
+  IERC20 public leapToken;
+  Bridge public bridge;
+  Vault public vault;
+
+  struct Proposal {
+    address initiator;
+    uint32 openTime;
+    bool finalized;
+    uint256 yesVotes;
+    uint256 noVotes;
+    mapping(address => int256) votes;
+    mapping(bytes32 => address) usedTxns;
+  }
+
+  mapping(bytes32 => Proposal) public proposals;
+
+  event ProposalRegistered(bytes32 indexed proposalHash, address indexed initiator);
+  event VoiceCasted(bytes32 indexed proposalHash, address indexed subject, uint256 weight, bytes32 txHash);
+  event VoiceChallenged(bytes32 indexed proposalHash, address indexed subject, address challenger);
+  event ProposalFinalized(bytes32 indexed proposalHash, bool isApproved);
+
+  constructor(address _mvgAddress, IERC20 _leapToken, Vault _vault) public {
+    mvgAddress = _mvgAddress;
+    leapToken = _leapToken;
+    vault = _vault;
+    bridge = _vault.bridge();
+  }
+
+  function _revertVote(Proposal storage proposal, bytes32 txHash) internal {
+    address signer = proposal.usedTxns[txHash];
+    delete proposal.usedTxns[txHash];
+    int256 votes = proposal.votes[signer];
+    delete proposal.votes[signer];
+    if (votes > 0) {
+      proposal.yesVotes -= uint256(votes);
+    } else {
+      proposal.noVotes -= uint256(votes * -1);
+    }
+  }
+
+  function registerProposal(bytes32 _proposalHash) public {
+    // make sure same proposals hasn't been opened before
+    require(proposals[_proposalHash].openTime == 0, "proposal already exists");
+    // get instance of token contract and pull proposal stake
+    leapToken.transferFrom(msg.sender, address(this), PROPOSAL_STAKE);
```

Comment:
I was wondering, what's gonna happen if initiator don't have enough LEAP? 🤔

Reply:
@iamonuwa and me think it will throw 😁